File: src/codes/scattered-instaniation/ts/rfc_singleton.ts
class Logger {
  private static instance: Logger;

  private constructor() {}

  static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }

  log(message: string) {
    console.log(`[LOG]: ${message}`);
  }
}

class UserService {
  private logger: Logger;

  constructor() {
    this.logger = Logger.getInstance(); // Singleton の取得
  }

  createUser(name: string) {
    this.logger.log(`ユーザー作成: ${name}`);
  }
}

// 利用例
const service = new UserService();
service.createUser("Hiroshi");

File: src/codes/scattered-instaniation/py/rfc_singleton.py
class Logger:
    _instance: "Logger" | None = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def log(self, message: str):
        print(f"[LOG]: {message}")

class UserService:
    def __init__(self):
        self.logger = Logger()  # 常に Singleton が返る

    def create_user(self, name: str):
        self.logger.log(f"ユーザー作成: {name}")

# 利用例
service = UserService()
service.create_user("Hiroshi")
